teps/0138-improve-tekton-pipeline-feature-flags.md

```diff
@@ -0,0 +1,73 @@
+
+### Feature stability coupled to CRD API version
+
+Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md), the behavior of `enable-api-fields` depends on the CRD API version being used. In v1beta1 CRDs, `beta` features can be enabled by setting `enable-api-fields` to `beta` or to "`stable`", but in v1 CRDs, `beta` features can only be enabled by setting `enable-api-fields` to `beta`. This couples API versioning to feature stability, and has led to the following pain points:
+
+- [Feedback indicates](https://github.com/tektoncd/pipeline/issues/6592#issuecomment-1533268522) that users upgrading their CRDs from v1beta1 to v1 were confused to find `beta` features that worked by default in v1beta1 did not work by default in v1 when `enable-api-fields` was set to "`stable`" (its default value). This is especially confusing for users who are not cluster operators and cannot control the value of `enable-api-fields`, especially if they are not aware they are using `beta` features.
+
+- For maintainers, the maintenance operation of swapping the storage version from v1beta1 to v1 should not have affected our users. However, we had to [change the user-facing default value of enable-api-fields from `stable` to `beta` ](https://github.com/tektoncd/pipeline/pull/6732) before changing the storage version of the API to [avoid breaking PipelineRuns using `beta` features](https://github.com/tektoncd/pipeline/pull/6444#issuecomment-1580926707).
+
+- When promoting features, it could cause confusions for contributors to be dependent on the fact whether an apiVersion is available. For example, during [the promotion to beta for projected workspaces](https://github.com/tektoncd/pipeline/pull/5530), the v1 api's existence led to confusions of what to do with beta features in v1beta1 and its difference with in v1.
+
+## Goals
+
+- [Decouple API versioning and feature stability](https://github.com/tektoncd/pipeline/issues/6592)
+  - Feature validation and implementation should be the same for all API versions.
+- Cluster operators can enable an individual feature without being forced to opt into other features at the same stability level.
+  - Increase the visibility of stability levels for features that have their own feature flags and ones that are not tied with the api spec
+- Describe a testing strategy that will give us confidence in our implementation
+  - One of the main reasons why [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-feature-gates.md#pros-and-cons) chose the existing style of feature flags was to have a simpler testing matrix.  With upgrades to the feature versioning, we need to ensure that our testing strategy gives us confidence in our feature coverage.
```

Comment:
NIT: This seems to me more of a requirement than a goal

Reply:
We have had similar ideas earlier and when coming up with requirements we figured that they might be pretty much overlapped (~80%) so we decide to merge the requirements and goal section since we might also want to do this as part of the goal of this proposal?